src/UnwindCursor.hpp

```diff
@@ -2697,7 +2697,13 @@ bool UnwindCursor<A, R>::setInfoForSigReturn(Registers_s390x &) {
   // own restorer function, though, or user-mode QEMU might write a trampoline
   // onto the stack.
   const pint_t pc = static_cast<pint_t>(this->getReg(UNW_REG_IP));
-  const uint16_t inst = _addressSpace.get16(pc);
+  uint16_t inst = 0;
+
+  struct iovec local_iov = {&inst, sizeof(inst)};
+  struct iovec remote_iov = {reinterpret_cast<void *>(pc), sizeof(inst)};
+  long bytesRead =
+      syscall(SYS_process_vm_readv, getpid(), &local_iov, 1, &remote_iov, 1, 0);
```

Comment:
This syscall is available since Linux 3.2, do we support earlier kernels in ClickHouse?

Reply:
Linux 3.2 was released on on 4 Jan, 2012, and this version is not in production any more.

Reply:
There was a similar function in libunwind:

DwarfInstructions.hpp:
```
// Check validity of the CFA.
// This is a very dirty hack (inspired by original libunwind version).
// Motivation: sometimes libunwind parse wrong value instead of CFA.
// We check that memory address belongs to our process by issuing "mincore" syscall.
// Actually we don't care if the address is in core or not, we only check return code.
// If Address Sanitizer will argue, replace syscall to inline assembly.
template <typename pint_t>
static bool isPointerValid(pint_t ptr)
{
  unsigned char mincore_res = 0;
  auto page_size = sysconf(_SC_PAGESIZE);
  return ptr && (0 == syscall(SYS_mincore, (void*)(ptr / page_size * page_size), 1, &mincore_res) || errno == ENOSYS);
}
```

Let's use it.

Reply:
@alexey-milovidov , unfortunately this function cannot filter out bad pointers as original fix. See the following segmentation fault after switching to using isPointerValid():
```
Thread 232 "QueryPullPipeEx" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x3fca7fbf840 (LWP 40705)]
libunwind::UnwindCursor<libunwind::LocalAddressSpace, libunwind::Registers_s390x>::setInfoForSigReturn (this=0x3fca7fb1848, this@entry=<error reading variable: value has been optimized out>) at /home/hlee/git/clib-z3/contrib/libunwind/src/UnwindCursor.hpp:2705
2705	  const uint16_t inst = _addressSpace.get16(pc);
(gdb) l 2705
2700	
2701	  if (!isPointerValid(pc)) {
2702	    return false;
2703	  }
2704	
2705	  const uint16_t inst = _addressSpace.get16(pc);
```
So I think we still need to keep the original fix.

Reply:
Just out of curiosity, @HarryLeeIBM do you have any guesses about why mincore does not work? Yes it is racy, but I doubt that this was the culprit.

Reply:
@azat, sorry I am not sure why and how bad unwind info was generated, but I am guessing it could generate some wild address values that cannot be filtered by sys_mincore system call.